### PR TITLE
Fix NPE

### PR DESCRIPTION
**Pull request by lmartelli**
Fixes https://github.com/webcompere/model-assert/issues/45

### CODE REVIEW

**Comment by ashleyfrieze** [comment]
Shall we add the test that characterises the bug?

I'm not sure that the null check ought to be done like this either. What if we had a rule that said "must be null" for the key?

Let's start with the tests that fail with the existing code, let's expand them for other edge conditions.

Then maybe we can quit the function early if there's no way the rule can pass.?

**Comment by lmartelli** [comment]
I've just added the test. Need help for a meaningful name ;-)

**Comment by lmartelli** [comment]
> Shall we add the test that characterises the bug?
> 
> I'm not sure that the null check ought to be done like this either. What if we had a rule that said "must be null" for the key?

Yes, you are right. I could not find where to put it elsewhere at first sight

**Comment by lmartelli** [comment]
It seems more appropriate to do the test in `PredicateWrappedCondition`, don't you think ?

**Comment by ashleyfrieze** [comment]
I've had a quick look at the code.

The comparison of specific number types is possible owing to the way that Jackson parses the numbers. We don't need to worry about that.

One fix might be in the `nodePasses` function of `NumberCondition`, which could reasonably do a null check on the input `jsonNode`. However, I'm starting to feel like the problem may be that there is no JsonNode.

File: src/test/java/uk/org/webcompere/modelassert/json/dsl/nodespecific/TreeComparisonDslTest.java [hunk 1]
```diff
@@ -175,4 +175,12 @@ void looseComparisonOfArraysWithContainsDoesntRequireOrderInSubTreeWithPathRule(
                 .path("a").arrayContains()
                 .isEqualTo("{a:[3, 2, 1]}"));
     }
+
+    @Test
+    void iAmNotInspiredToFindANameForThisTest() {
```
Comment (ashleyfrieze) [inline src/test/java/uk/org/webcompere/modelassert/json/dsl/nodespecific/TreeComparisonDslTest.java:180]: `givenObjectWithMissingKeyThenCanStillHaveRulesForWhenKeyIsPresent`

**Comment by ashleyfrieze** [comment]
@lmartelli - thanks for trying to fix this. I think you're right that the fix belongs in `PredicateWrappedCondition`. Please rename the test, and then I'll merge this fix and release it.

**Comment by ashleyfrieze** [comment]
The various numeric conditions are a bonus. Though JSON doesn't have more than `number` as a type, Java does, and we can still assert the convertability to other formats of the number, so there's no problem with `isInteger` etc.

**Comment by lmartelli** [comment]
> The various numeric conditions are a bonus. Though JSON doesn't have more than `number` as a type, Java does, and we can still assert the convertability to other formats of the number, so there's no problem with `isInteger` etc.

My point is that it is currently a little cumbersome if you do not care about the Java type, which is a bit odd in a JSON context. Typical use of JSON is web API, where Java type are irrelevant. 

Also, `isNumber()` is too broad : you usually want to distinguish integers and decimal numbers.

How about adding `isIntegralNumber()` ?

**Review by ashleyfrieze — APPROVED** [review]

**Comment by ashleyfrieze** [comment]
@lmartelli  - please raise a fresh PR if you wish to try adding more variants of type detection.

From `JsonNode`:

```java
    public final boolean isPojo() {
        return this.getNodeType() == JsonNodeType.POJO;
    }

    public final boolean isNumber() {
        return this.getNodeType() == JsonNodeType.NUMBER;
    }

    public boolean isIntegralNumber() {
        return false;
    }

    public boolean isFloatingPointNumber() {
        return false;
    }

    public boolean isShort() {
        return false;
    }

    public boolean isInt() {
        return false;
    }

    public boolean isLong() {
        return false;
    }

    public boolean isFloat() {
        return false;
    }

    public boolean isDouble() {
        return false;
    }

    public boolean isBigDecimal() {
        return false;
    }

    public boolean isBigInteger() {
        return false;
    }

    public final boolean isTextual() {
        return this.getNodeType() == JsonNodeType.STRING;
    }

    public final boolean isBoolean() {
        return this.getNodeType() == JsonNodeType.BOOLEAN;
    }

    public final boolean isNull() {
        return this.getNodeType() == JsonNodeType.NULL;
    }

    public final boolean isBinary() {
        return this.getNodeType() == JsonNodeType.BINARY;
    }
```

So, theoretically, we could provide examples of all of these. It may be better to write tests for any ones you wish to expose inside of `WholeTreeUseCaseTest.java` rather than the file you've added tests for in this issue.

**Comment by ashleyfrieze** [comment]
@lmartelli  - I've triggered release 1.0.2 - takes a few hours to propagate to your favourite maven repo.

**Comment by lmartelli** [comment]
> @lmartelli - I've triggered release 1.0.2 - takes a few hours to propagate to your favourite maven repo.

Thanks, happy new year 🥳
